cpp/src/community/detail/common_methods.cuh

```diff
@@ -0,0 +1,370 @@
+                    static_cast<vertex_t const*>(numbering_indices.begin())),
+    new_graph_view.local_vertex_partition_range_size(),
+    labels.data(),
+    labels.size(),
+    false);
+
+  return std::move(new_graph);
+}
+
+template <typename graph_view_t>
+void update_by_delta_modularity(
+  raft::handle_t const& handle,
+  graph_view_t const& graph_view,
+  typename graph_view_t::weight_type total_edge_weight,
+  typename graph_view_t::weight_type resolution,
+  rmm::device_uvector<typename graph_view_t::weight_type>& vertex_weights_v,
+  rmm::device_uvector<typename graph_view_t::vertex_type>& cluster_keys_v,
+  rmm::device_uvector<typename graph_view_t::weight_type>& cluster_weights_v,
+  rmm::device_uvector<typename graph_view_t::vertex_type>& next_clusters_v,
+  edge_src_property_t<graph_view_t, typename graph_view_t::weight_type> const&
+    src_vertex_weights_cache,
+  edge_src_property_t<graph_view_t, typename graph_view_t::vertex_type>& src_clusters_cache,
+  edge_dst_property_t<graph_view_t, typename graph_view_t::vertex_type>& dst_clusters_cache,
+  bool up_down)
+{
+  using vertex_t = typename graph_view_t::vertex_type;
+  using weight_t = typename graph_view_t::weight_type;
+
+  rmm::device_uvector<weight_t> vertex_cluster_weights_v(0, handle.get_stream());
+  edge_src_property_t<graph_view_t, weight_t> src_cluster_weights(handle);
+  if constexpr (graph_view_t::is_multi_gpu) {
+    // TODO:  Shouldn't this be int_vertex_t?
+    cugraph::detail::compute_gpu_id_from_ext_vertex_t<vertex_t> vertex_to_gpu_id_op{
+      handle.get_comms().get_size()};
+
+    vertex_cluster_weights_v =
+      cugraph::collect_values_for_keys(handle.get_comms(),
+                                       cluster_keys_v.begin(),
+                                       cluster_keys_v.end(),
+                                       cluster_weights_v.data(),
+                                       next_clusters_v.begin(),
+                                       next_clusters_v.end(),
+                                       vertex_to_gpu_id_op,
+                                       invalid_vertex_id<vertex_t>::value,
+                                       std::numeric_limits<weight_t>::max(),
+                                       handle.get_stream());
+
+    src_cluster_weights = edge_src_property_t<graph_view_t, weight_t>(handle, graph_view);
+    update_edge_src_property(
+      handle, graph_view, vertex_cluster_weights_v.begin(), src_cluster_weights);
+    vertex_cluster_weights_v.resize(0, handle.get_stream());
+    vertex_cluster_weights_v.shrink_to_fit(handle.get_stream());
+  } else {
+    thrust::sort_by_key(handle.get_thrust_policy(),
+                        cluster_keys_v.begin(),
+                        cluster_keys_v.end(),
+                        cluster_weights_v.begin());
+
+    vertex_cluster_weights_v.resize(next_clusters_v.size(), handle.get_stream());
+    thrust::transform(handle.get_thrust_policy(),
```

Comment:
// for each cluster found in next_clusters_v_, lookup its weight in cluster_weights_v_
    // and store them in local variable vertex_cluster_weights_v

Reply:
Done in next push